### PR TITLE
dropping variables from a sampleset

### PR DESCRIPTION
**Pull request by pau557**
It simply removes variables from a sampleset

### CODE REVIEW

File: tests/test_sampleset.py [hunk 1]
```diff
@@ -489,6 +489,16 @@ def test_invalid(self):
...
+class TestDropVariables(unittest.TestCase):
+    def test_variables_were_dropped(self):
+        samples = np.triu(np.ones((5, 5)))
+        labels = 'abcde'
+        sampleset = dimod.SampleSet.from_samples((samples, labels), vartype='BINARY', energy=1)
+        variables_to_drop="ae"
+        expected_variables = set([label for label in labels if label not in variables_to_drop])
+        obtained_variables = set(dimod.sampleset.drop_variables(sampleset, variables_to_drop).variables)
+        self.assertTrue(obtained_variables==expected_variables)
+
```
Comment (arcondello) [inline tests/test_sampleset.py:501]: Can we also check that the correct sample values were returned?

File: tests/test_sampleset.py [hunk 2]
```diff
@@ -489,6 +489,16 @@ def test_invalid(self):
         with self.assertRaises(ValueError):
             sampleset = dimod.append_data_vectors(sampleset, sets=[{0}, {1}])
 
+class TestDropVariables(unittest.TestCase):
+    def test_variables_were_dropped(self):
+        samples = np.triu(np.ones((5, 5)))
+        labels = 'abcde'
+        sampleset = dimod.SampleSet.from_samples((samples, labels), vartype='BINARY', energy=1)
```
Comment (arcondello) [inline tests/test_sampleset.py:496]: Also, need to either import `dimod` at the top or just drop the `dimod.` here.

File: dimod/sampleset.py [hunk 3]
```diff
@@ -453,6 +453,27 @@ def infer_vartype(samples_like):
...
     raise ValueError("given samples_like is of an unknown vartype")
 
+def drop_variables(sampleset, variables_to_drop):
+    """Drop variables from a sampleset
+
+    Args:
+        sampleset (:obj:`.SampleSet`):
+            Original sampleset.
+
+        variables_to_drop (iter):
```
Comment (arcondello) [inline dimod/sampleset.py:463]: should be `container`, not `iter`. An iterator won't actually work in this context.

File: dimod/sampleset.py [hunk 4]
```diff
@@ -453,6 +453,27 @@ def infer_vartype(samples_like):
 
     raise ValueError("given samples_like is of an unknown vartype")
 
+def drop_variables(sampleset, variables_to_drop):
+    """Drop variables from a sampleset
+
+    Args:
+        sampleset (:obj:`.SampleSet`):
```
Comment (randomir) [inline dimod/sampleset.py:460]: One should [use `class`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cross-referencing-python-objects) for a specific type.
```suggestion
        sampleset (:class:`.SampleSet`):
```

File: dimod/sampleset.py [hunk 5]
```diff
@@ -453,6 +453,27 @@ def infer_vartype(samples_like):
 
     raise ValueError("given samples_like is of an unknown vartype")
 
+def drop_variables(sampleset, variables_to_drop):
+    """Drop variables from a sampleset
```
Comment (randomir) [inline dimod/sampleset.py:457]: We should be clear we are returning _a new_ sampleset (a copy), not modifying the original one.

File: dimod/sampleset.py [hunk 6]
```diff
@@ -453,6 +453,27 @@ def infer_vartype(samples_like):
...
+
+    Args:
+        sampleset (:obj:`.SampleSet`):
+            Original sampleset.
+
+        variables_to_drop (iter):
+            Iterable containing the variables to be dropped
+
+    Returns:
+        :obj:`.SampleSet`: A sampleset without the variables to be dropped.
```
Comment (randomir) [inline dimod/sampleset.py:467]:
```suggestion
        :class:`.SampleSet`: A sampleset without the variables to be dropped.
```

File: dimod/sampleset.py [hunk 7]
```diff
@@ -453,6 +453,27 @@ def infer_vartype(samples_like):
...
+    """
+    
+    reduced_variables = [var for var in sampleset.variables if var not in variables_to_drop]
+    reduced_samples = sampleset.samples()[:, reduced_variables]
+
+    return dimod.SampleSet.from_samples((reduced_samples, reduced_variables), 
+                                        vartype=sampleset.vartype, 
+                                        energy=sampleset.record.energy,
+                                        info=sampleset.info)
```
Comment (randomir) [inline dimod/sampleset.py:476]: Should make a (deep)copy of the original info dict.

File: dimod/sampleset.py [hunk 8]
```diff
@@ -453,6 +453,27 @@ def infer_vartype(samples_like):
...
+            Original sampleset.
+
+        variables_to_drop (iter):
+            Iterable containing the variables to be dropped
+
+    Returns:
+        :obj:`.SampleSet`: A sampleset without the variables to be dropped.
+    """
+    
+    reduced_variables = [var for var in sampleset.variables if var not in variables_to_drop]
```
Comment (randomir) [inline dimod/sampleset.py:470]: I'd "settify" `variables_to_drop` before doing N existence queries.

File: dimod/sampleset.py (hunk 3 above)
Comment (randomir) [inline dimod/sampleset.py:463]: Consider renaming to `variables`. Should be clear from the context, and it's easier on the eyes.

File: dimod/sampleset.py [hunk 9]
```diff
@@ -450,6 +459,58 @@ def infer_vartype(samples_like):
...
+    return keep_variables(sampleset, sampleset.variables - variables)
+
+
+def keep_variables(sampleset: 'SampleSet', variables: Iterable[Variable]) -> 'SampleSet':
+    """Return a new sample set with only the given variables.
+
+    Args:
+        sampleset: A sample set.
+        variables: The variables to be kept. Must be a subset of the variables
+            in the sample set.
```
Comment (randomir) [inline dimod/sampleset.py:484]: For symmetry with `drop_variables`, why not allow variables not in the sampleset?

**Review by randomir — APPROVED** [review]
LGTM

File: dimod/sampleset.py (hunk 9 above)
Comment (arcondello) [reply dimod/sampleset.py:484]: In this case the asymmetry is deliberate. I think of it as "drop any of these variables" and "keep all of these variables". So if I was able to provide variables not in the sample set to keep, I would expect the created sample set to have all of the ones specified.

Comment (randomir) [reply dimod/sampleset.py:484]: Sure, I get that. Personally, I would appreciate consistency, so at least for drop to require a subset... But I really don't have strong feeling about it. :shrug:

**Pull request merged**